### DISCUSSION TITLE
tutorials/lift-off-part2/06-implementing-query-resolvers

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#34] [course feedback]
# tutorials/lift-off-part2/06-implementing-query-resolvers

https://www.apollographql.com/tutorials/lift-off-part2/06-implementing-query-resolvers

**Comment by GeroWalther** [comment]
How do I know that I can access to authorId on the parent argument in the resolver function? How to set another name if needed?

**Reply by mabuyo** [reply]
Hey @GeroWalther! You can always try a `console.log(parent)` in the resolver function to get the full details of what the `parent` returns. In this example in the lesson, the resolvers will run like this:

- `Query.tracksForHome()` -> returns a list of `Track` objects, as returned by the RESTDataSource `GET /tracks` method (so you can look at the [REST API documentation](https://odyssey-lift-off-rest-api.herokuapp.com/docs/#/Tracks/get_tracks) to see what that endpoint returns)
- Then, `Track.author()` for each track in the list above. The `parent` argument is going to contain the track object. In that track object, we have access to the `authorId` property.

Can you explain your second question about setting another name? What name are you referring to?

**Reply by reactivicky** [reply]
Thank you. This lesson took some re-reading for me to understand.

**Comment by aniketicloud** [comment]
- I was confused second time, both in version 1 and version 2 about authorId.
- I didn't understand from where authorId was coming from. I thought it should have defined in schema, but there was not. 
- Our server is not exposing authorId field coming from rest api response.

**Reply by mabuyo** [reply]
Hey @aniketicloud! You can always try a `console.log(parent)` in the resolver function to get the full details of what the `parent` returns. In this example in the lesson, the resolvers will run like this:

- `Query.tracksForHome()` -> returns a list of `Track` objects, as returned by the RESTDataSource `GET /tracks` method (so you can look at the [REST API documentation](https://odyssey-lift-off-rest-api.herokuapp.com/docs/#/Tracks/get_tracks) to see what that endpoint returns)
- Then, `Track.author()` for each track in the list above. The `parent` argument is going to contain the track object. In that track object, we have access to the `authorId` property.

And to further clarify, `authorId` is _not_ defined in the schema, but it is available through our REST datasource. However, you can still retrieve that information by building your query like this:

```graphql
query GetTracksAuthors {
  tracksForHome {
    title
    author {
      id # this contains the same value as the authorId
    }
  }
```

**Reply by Jawad-repo** [reply]
@mabuyo in this part of course, the final example code that is just above the last Code Challenge is missing export at the end i.e module.exports = resolvers;

**Reply by lizhennessy** [reply]
Thanks @Jawad-repo ! We'll update this.

**Comment by mafarazuddin** [comment]
I am confused with this :) 

Why my answer is wrong here - what has name to do with my resolver definition?

/* your code:
const resolvers = {
  Query: {
    spaceCats: (_, __, {dataSources}) => {
      return dataSources.spaceCatsAPI.getSpaceCats();
    }
  },

  Mission: {

    mission:({catId},_,{dataSources}) => {
      return dataSources.spaceCatsAPI.getMissions(catId);
    },
  },
  // write the SpaceCat.missions resolver below




};
*/

**Reply by mabuyo** [reply]
Hey @mafarazuddin ! It looks like your resolver is under the `Mission` key, not the `SpaceCat` key. The resolver object map should match the schema. A `SpaceCat` type has a `missions` field, so the resolver function should be under `SpaceCat.mission`, _not_ `Mission.mission` (as your code shows).

Let me know if I can clarify further!

**Comment by Hydrock** [comment]
Why do you force to follow the convention with underscores. This only complicates the lesson and has nothing to do with GraphQL at all- well, variables are not used, and well and good. Why complicate it?

**Comment by Hydrock** [comment]
The lesson is just terrible. You started so well, but now everything is bad.

I am already an experienced developer - he misled me during the lesson and made me nervous and think that I don't understand something.

On Lesson start we wrote "return dataSources.trackAPI.getTracksForHome(); " There is even a note in the article: (lowercase here as it's the instance of our TrackAPI class extending RESTDataSource)

I immediately started thinking that I missed creating an instance of the class - I returned to the last lesson and found nothing - then I started thinking that an instance of the class creates GraphQL itself magically and we need to observe a special structure in the file system.

And only at the very end do I read this:

Now, you might be wondering where the dataSources object that's passed to our resolver comes from. When did we add it to the contextValue parameter? Right now, it isn't added. If we ran our server, it would have no clue about our TrackAPI. It also doesn't know about our resolvers yet! Let's make sure all three pieces are connected with each other.

This is not how the code is written. It's very strange. We had to first create an instance of the class, provide it to graphml, and only then write methods using it. Yes, this can be done in real development when you yourself know what you're doing, but here I can't predict the author's thoughts.

I'm sorry for the criticism - but it made me mad

**Comment by edward-hunt** [comment]


**Reply by lizhennessy** [reply]
Hi there @edward-hunt ! You're on the right track - we definitely have that inconsistency to overcome. In an [upcoming lesson on codegen](https://www.apollographql.com/tutorials/lift-off-part2/08-server-codegen), we'll tackle this problem by defining a custom `TrackModel` that lets us bridge the gap. Could you screenshot the exact error that you're seeing? I'll be happy to take a closer look if it's blocking your progress on the course!

**Reply by edward-hunt** [reply]
The link you provided solved all my problems! I knew there would be a codegen related solution and did not realize there was a JS and a TS option for this course =) Thank you!

**Comment by Deepak22448** [comment]
I am using TypeScript , what are the types for parent , args and context ? help me with this.

**Reply by lizhennessy** [reply]
Hi there @Deepak22448 ! Thanks for reaching out — you're absolutely on the right track with your question. There's a lesson later in this course called "Codegen on the server", and in it we'll see how we use the GraphQL Codegen package to generate all the types for our resolver arguments. If you're looking to do this independent of the codegen package, you can check out this [community post](https://community.apollographql.com/t/typescript-types-for-resolvers/5272) to read up on how our maintainers have stepped through the process in the past!

**Reply by Deepak22448** [reply]
@lizhennessy thanks a lot , you guys are doing great for newbies like me.

**Comment by neogeek13** [comment]
It took me a long time to realize that the defined parent (authorID) is expected to be CONTAINED BY the parent and not BE the parent (authorID as a parent to... the tracks? didn't make sense).

I had returned to the prior lesson to review the definition of 'parent' there and it didn't illuminate the intent any more clearly:

"parent:
parent is the returned value of the resolver for this field's parent. This will be useful when dealing with resolver chains."

This definition seems circular in that it references itself ('parent') in the definition. If you still don't understand what 'parent' is in this context, that definition is more confusing than helpful. I glossed over it at the time assuming I'd know what it was as I understand the hierarchical concepts of parent and child in many other contexts.

What finally made it click for me was one of the answers above (others appear to be having various challenges around the authorID/parent piece here as well tho possibly for different reasons):
"The parent argument is going to contain the track object. In that track object, we have access to the authorId property."

Specifically "The parent argument is going to contain the track object."

So perhaps somehow stating up front that the parent will list 'a field' that is expected 'to be contained within the object' returned for its use? 

I still wasn't sure how "The parent argument contains data returned by our tracksForHome resolver," was tied to anything until I saw:
"Our resolvers object's keys will correspond to our schema's types and fields."

This understated sentence is the, if you'll pardon the pun, 'key' to the whole thing. This deserves a few sentences on its own to tie 'Query' and 'Track' back to the mock and the API after which it all falls into place. If I had understood the significance of the key in resolvers I would have seen the connections and you wouldn't be reading this. :P

**Reply by neogeek13** [reply]
The "Codegen on the server: Adding a Model type" section adds still more important color on this. I certainly understand that you have to pick and choose an order in which to present information. This gap in explaining more about how the parent functions in the larger scope of things took a lot of my time as I tried to piece together what I thought I was certainly just missing (vs. what hadn't been presented thoroughly for someone new to the topic).

Stating that "we could pluck the authorId field from each track object to call our datasource for information about a particular author." confirms the concept that what is returned for the parent is in fact NOT the authorID but data from which authorID can be 'plucked' - somewhat unusual in my experience of how arguments are typically used (you pass the specific value that is requested). 

Please let me know if I'm completely missing the point here - I feel pretty certain I'm grokking how this aspect works now but would certainly like to know if I missed something.

**Reply by neogeek13** [reply]
Annnnd the Course 3: "Resolver chains" lesson was really what I needed to see. In some ways I'd almost prefer "Hey, we'll cover this for real in a future lesson - for now just trust us that it works and go ahead and use it."

**Comment by davidemmolino** [comment]
How does Apollo server know to iterate over the tracks list when we call the author resolves? Does this behavior happen for each resolver outside of "Query"?

**Reply by lizhennessy** [reply]
Hey there, @davidemmolino! Great question.

When we run a query like the one below, we'll automatically trigger the `Query.tracksForHome` resolver that we've defined in our `resolvers` file. That particular resolver function, as we've seen in this lesson, returns an array of track objects, and we've indicated in our schema that each of those objects should satisfy the `Track` type. This means that a query like the below - which requests the `title` field for each returned track - will read each of the track objects we get from our REST API and do its best to return the `title` field for _each_. (It's already iterating over the list of tracks!) 

```
query GetTracksTitles {
  tracksForHome {
    title
  }
}
```

It gets a bit more interesting when we add something like fields for a track's author:

```
query GetTracksAuthors {
     tracksForHome {
         title
         author {
            id
            name
            photo
         }
     }
}
```

Because the `Track.author` field is denoted in our schema as returning an `Author` type, you can kind of think of this as a portal we use to "jump" into the selection of fields on the `Author` type. Now when we run this query, our server takes in the list of track objects, extracts out their `title` field to return, and then encounters the `author` field in the query. The objects we get back from the REST API do not have an `author` field, but they _do_ have an `authorId` field-- so this is where our `Track.author` field comes in! In it, we write all the instructions the server needs to fulfill this field for _each_ of the track objects it's returning. 

I hope that helps clarify a bit what's happening under the hood, but of course please reach out with any other questions. You'll see this in a lot more detail in the next course, which goes deeper into [resolver chains](https://www.apollographql.com/tutorials/lift-off-part3/05-resolver-chains)!

**Comment by mandyparson** [comment]
@mabuyo thanks for all your help here! Just a note: console.log(parent) doesn't work at this point in the lesson because we are using mock data. We haven't linked up the resolvers yet. It works after doing the lesson after this!

**Reply by repiemm** [reply]
Now I see, this is the comment I needed to read before going crazy because the console.log() was not working haha, thanks a lot.

**Comment by BlockBusterk** [comment]
Does the resolver name has to match the field's name in the schema graphql

**Reply by mabuyo** [reply]
Yes, that's the way it works in Apollo Server!

**Comment by ryany08** [comment]
Does it mean that if I don't add field id in type author to query string tracksForHome, the authorId in parent wont't work?

**Reply by ryany08** [reply]
I just realized that the authorId returned from REST API will correspond to the author in query string. So I guess it doesn't matter?

**Reply by lizhennessy** [reply]
Hi there @ryany08  - great question. It can be tempting with GraphQL to make our schema a 1:1 match with our REST API, but we'll see pretty quickly that it's actually not necessary to do so. For example, as you saw, the REST API response from `https://odyssey-lift-off-rest-api.herokuapp.com/tracks` returns a list of tracks, each of which has an `authorId`. When we want to make a follow-up request to get _more_ details about an author's track, we'll use this property in our next REST API request. The value that we _return_ from our `Track.author` resolver should match the shape that we defined for an `Author` type: but what that actually looks like is up to us!

So currently, we take the `authorId` for each track, and make a request to the REST API for that particular author's details.

```
author: ({ authorId }, _, { dataSources }) => {
      return dataSources.trackAPI.getAuthor(authorId);
    },
```

If we look at this response (going to [this link](https://odyssey-lift-off-rest-api.herokuapp.com/author/cat-1) for example) we get an object back in this shape.

```
{
  "id": "cat-1",
  "name": "Henri, le Chat Noir",
  "photo": "https://images.unsplash.com/photo-1442291928580-fb5d0856a8f1?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjExNzA0OH0"
}
```

So we can see that our REST response includes the properties `id`, `name`, and `photo`. And those are the properties that we've given to our `Author` type! 

```
"Author of a complete Track or a Module"
type Author {
  id: ID!
  name: String!
  photo: String
}
```

What if the properties returned by the REST API _didn't_ match our `Author` type? For example, maybe instead of `photo` we get back `photoThumbnailUrl`. Well, that's okay - we don't need to change the `Author` type in our schema to match. We might think that `photo` is a more intuitive and terse property, and we want the querying process to be easy for our frontend consumers. In that case, we could define a resolver specific to the `Author.photo` field.

```
Author: {
    // what should I return when an author's photo is queried?
    photo: () => {}
}
```

We could use _this_ resolver to access the object we get from the REST API (with that hefty `photoThumbnailUrl` field) as `parent`. From there we'd pluck off the property we _want_ to be returned as `photo`... and return it!

```
Author: {
    photo: ({ photoThumbnailUrl }) => {
       return photoThumbnailUrl;
    }
}
```

In this way, we get really specific, fine-grained control over the values we get to return for each field in our schema. And this works even when the REST API differs from the shape we want our schema to take! Hopefully this explanation wasn't too long. Please reach out if you have any other questions!

**Comment by MohammadEzz** [comment]
is this way for querying data using parent resolver and it's child resolver fix the problem of N+1 requests

**Reply by lizhennessy** [reply]
hey @MohammadEzz ! Great question. Using field-specific resolvers can definitely help us with the _amount_ of data we end up requesting to fulfill certain parts of our schema, however the N+1 problem is still something we'll run into. Frequently this occurs when we have one resolver function that we need to execute many many times for a _list_ of objects. Instead, the goal is to batch _all_ of those requests into one, so that we're not pinging the same endpoint with a different ID each time.

We've got a separate course all about [Data loaders with TypeScript](https://www.apollographql.com/tutorials/dataloaders-typescript/03-data-loaders-under-the-hood) that goes into how we solve this problem, and how it all works under the hood! (And the same for [Java/DGS](https://www.apollographql.com/tutorials/dataloaders-dgs), if that appeals more.)
